core/base/dim.hpp

```diff
@@ -0,0 +1,229 @@
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_BASE_DIM_HPP_
+#define GKO_CORE_BASE_DIM_HPP_
+
+
+#include "core/base/types.hpp"
+#include "core/base/utils.hpp"
+
+
+namespace gko {
+
+
+/**
+ * A type representing the dimensions of a multidimensional object.
+ *
+ * @tparam Dimensionality  number of dimensions of the object
+ * @tparam DimensionType  datatype used to represent each dimension
+ */
+template <size_type Dimensionality, typename DimensionType = size_type>
+struct dim {
+    static constexpr size_type dimensionality = Dimensionality;
+
+    using dimension_type = DimensionType;
+
+    /**
+     * Creates a dimension object with all dimensions set to the same value.
+     *
+     * @param size  the size of each dimension
+     */
+    constexpr GKO_ATTRIBUTES dim(const dimension_type &size = dimension_type{})
+        : first_{size}, rest_{size}
+    {}
+
+    /**
+     * Creates a dimension object with the specified dimensions.
+     *
+     * If the number of dimensions given is less than the dimensionality of the
+     * object, the remaining dimensions are set to the same value as the last
+     * value given.
+     *
+     * @param first  first dimension
+     * @param rest  other dimensions
+     */
+    template <typename... Rest>
+    constexpr GKO_ATTRIBUTES dim(const dimension_type &first,
+                                 const Rest &... rest)
+        : first_{first}, rest_{static_cast<dimension_type>(rest)...}
+    {}
+
+    /**
+     * Returns the requested dimension.
+     *
+     * @param dimension  the requested dimension
+     *
+     * @return the `dimension`-th dimension
+     */
+    constexpr GKO_ATTRIBUTES const dimension_type &operator[](
+        const size_type &dimension) const noexcept
+    {
+        return GKO_ASSERT(dimension < dimensionality), *(&first_ + dimension);
+    }
+
+    /**
+     * @copydoc operator[]() const
+     */
+    GKO_ATTRIBUTES dimension_type &operator[](
+        const size_type &dimension) noexcept
+    {
+        return GKO_ASSERT(dimension < dimensionality), *(&first_ + dimension);
+    }
+
+    /**
+     * Checks if all dimensions evaluate to true.
+     *
+     * For standard arithmetic types, this is equivalent to all dimensions being
+     * different than zero.
+     *
+     * @return true if and only if all dimensions evaluate to true
+     */
+    constexpr GKO_ATTRIBUTES operator bool() const
+    {
+        return static_cast<bool>(first_) && static_cast<bool>(rest_);
+    }
+
+    /**
+     * Checks if two dime objects are equal.
+     *
```

Comment:
Could something like "subset" be reasonable? like, each dimensionality being smaller / equal the dimensionality of the other object. Just an idea...

Reply:
Maybe, it's not at this point so I didn't add it - I only implemented utilities that  were needed somewhere else.